# Leicester v. Wolverhampton Wanderers - 2024-12-22 match analysis
*This notebook intends to walk through the theory and calculations I am currently using in my model to calculate probability of 1 x 2 result and a table of over / under goal projections. I will break the calculations out into utility functions stored in `../src/utils` for ease of use in future explorations. This is to document the calculations.*

### Instructions:

1. Ensure all cells in all notebooks in `/notebooks` directory have been run.
2. Define `home_side` and `away_side` in first cell. If making a comparison with odds from `03_retrieve_odds.ipynb` also enter matchday.
3. Click 'run all'
4. Probability for 1 x 2 and over / under can be viewed in the last cell
5. *Suggestion:* copy and paste notebook to `../output/< matchday_directory >/` and rename file. IMPORTANT: Update file path in lines 14 and 15 if needed if you make a copy.



### Import League Data & Define Sides

Define 'home_side' and 'away_side'
Ensure correct output



In [19]:
import pandas as pd
from scipy.stats import poisson
import numpy as np

"""
BUG: if copying from `notebooks/03_analysis_template_notebook.ipynb` to a directory in output
for storage, ensure the file path for reading the csv will need to be updated from '../data/processed/home_table.csv'
to '../../data/processed/home_table.csv' if frames are run again.

TLDR
if you are in ./notebooks/03_analysis_tempalte.csv you may run this cell as is. If you are not, check that following
read_csv commands read from the correct directory.
"""
home_dataframe = pd.read_csv('../data/processed/home_table.csv')
away_dataframe = pd.read_csv('../data/processed/away_table.csv')

"""
!!! INPUT TEAMS & MATCHDAY HERE !!!

wrapped in quotes, eg.

home_side = "Bournemouth"
away_side = "West Ham"
matchday = "2024-12-16" # (optional)

If there is a ValueError,
(1) check spelling against the error message. Must be exact.
(2) check capitalization
(3) check that it is wrapped in quotes

!!! INPUT TEAMS & MATCHDAY HERE !!!
"""
home_side = "Leicester"
away_side = "Wolverhampton Wanderers"
matchday = "2024-12-22" # (optional)

# validate input
clubs = set(home_dataframe['Team'].values)
if home_side not in clubs:
    raise ValueError(f"Invalid entry for 'home_side'. Expected one of {clubs}, but received '{home_side}'")
elif away_side not in clubs:
    raise ValueError(f"Invalid entry for 'away_side'. Expected one of {clubs}, but received '{away_side}'")
else:
    print(f"Ready to calculate probability for {home_side} v. {away_side}")
    print("\nNOTEBOOK TITLE (copy and paste): \n")
    print(f"{home_side} v. {away_side} - {matchday} match analysis")

Ready to calculate probability for Leicester v. Wolverhampton Wanderers

NOTEBOOK TITLE (copy and paste): 

Leicester v. Wolverhampton Wanderers - 2024-12-22 match analysis


### Calculate Projected Goals

In [20]:
# 2. grab their gpm_scored and gpm_conceded
home_side_gpm_scored = home_dataframe.loc[
    home_dataframe["Team"] == home_side,
    "gpm_scored"
].values[0]

home_side_gpm_conceded = home_dataframe.loc[
    home_dataframe["Team"] == home_side,
    "gpm_conceded"
].values[0]

print(home_side_gpm_scored, home_side_gpm_conceded)

away_side_gpm_scored = away_dataframe.loc[
    away_dataframe["Team"] == away_side,
    "gpm_scored"
].values[0]

away_side_gpm_conceded = away_dataframe.loc[
    away_dataframe["Team"] == away_side,
    "gpm_conceded"
].values[0]

print(away_side_gpm_scored, away_side_gpm_conceded)

# 3.
epl_home_avg_gpm_scored = round(sum(home_dataframe["gpm_scored"].values) / 20, 2)
epl_home_avg_gpm_conceded = round(sum(home_dataframe["gpm_conceded"].values) / 20, 2)
print(epl_home_avg_gpm_scored, epl_home_avg_gpm_conceded)

# 4. create a ratio (attack and defense rating) of their gpm to the league average gpm
home_attack_rating = home_side_gpm_scored / epl_home_avg_gpm_scored
away_defense_rating = away_side_gpm_conceded / epl_home_avg_gpm_scored
away_attack_rating = away_side_gpm_scored / epl_home_avg_gpm_conceded
home_defense_rating = home_side_gpm_conceded / epl_home_avg_gpm_conceded

print(home_attack_rating, away_defense_rating, away_attack_rating, home_defense_rating)

# home projected goals
home_projected_goals = home_attack_rating * away_defense_rating * epl_home_avg_gpm_scored
# away projected goals
away_projected_goals = away_attack_rating * home_defense_rating * epl_home_avg_gpm_conceded
print("\n ================== projected_goals ==================== \n")
print(f"HOME {home_side} projected goals:", home_projected_goals)
print(f"AWAY {away_side} projected goals: ", away_projected_goals)

1.375 1.5
1.5 2.5
1.61 1.4
0.8540372670807453 1.5527950310559004 1.0714285714285714 1.0714285714285714

 ================== projected_goals ==================== 

HOME Leicester projected goals: 2.135093167701863
AWAY Wolverhampton Wanderers projected goals:  1.607142857142857


### Poisson Distribution of Goal Probabilities

In [21]:
# Poisson probability for home team goals 0 - 8
home_score_prob = []
away_score_prob = []
for i in range(9):
    prob = poisson.pmf(i, home_projected_goals)
    home_score_prob.append(prob)
for i in range(9):
    prob = poisson.pmf(i, away_projected_goals)
    away_score_prob.append(prob)

home_score_prob = np.array(home_score_prob)
away_score_prob = np.array(away_score_prob)
# print(home_score_prob)
# print(away_score_prob)

exact_score_prob = np.outer(away_score_prob, home_score_prob)
print(f"{home_side} v. {away_side}", exact_score_prob)

Leicester v. Wolverhampton Wanderers [[2.37010477e-02 5.06039450e-02 5.40220686e-02 3.84473832e-02
  2.05221863e-02 8.76335595e-03 3.11843024e-03 9.51162728e-04
  2.53852630e-04]
 [3.80909695e-02 8.13277688e-02 8.68211817e-02 6.17904373e-02
  3.29820851e-02 1.40839649e-02 5.01176288e-03 1.52865438e-03
  4.07977441e-04]
 [3.06088148e-02 6.53526713e-02 6.97670210e-02 4.96530300e-02
  2.65034613e-02 1.13174718e-02 4.02730946e-03 1.22838299e-03
  3.27839015e-04]
 [1.63975794e-02 3.50103596e-02 3.73751898e-02 2.65998375e-02
  1.41982828e-02 6.06293133e-03 2.15748721e-03 6.58062314e-04
  1.75628044e-04]
 [6.58831313e-03 1.40666624e-02 1.50168173e-02 1.06874347e-02
  5.70466721e-03 2.43599919e-03 8.66847540e-04 2.64400037e-04
  7.05648391e-05]
 [2.11767208e-03 4.52142719e-03 4.82683415e-03 3.43524687e-03
  1.83364303e-03 7.82999741e-04 2.78629566e-04 8.49857262e-05
  2.26815554e-05]
 [5.67233593e-04 1.21109657e-03 1.29290200e-03 9.20155412e-04
  4.91154383e-04 2.09732074e-04 7.46329195e-05 2.

### Under Market

In [22]:
thresholds = [0.5, 1.5, 2.5, 3.5]
sums = []

for t in thresholds:
    mask = np.add.outer(
        np.arange(
            exact_score_prob.shape[0]
        ),
        np.arange(
            exact_score_prob.shape[1]
        )
    ) <= t
    # print(mask)
    sums.append(exact_score_prob[mask].sum())

# print(sums)
under_market_df = pd.DataFrame({
    "Goals": thresholds,
    "Prob": sums
})
under_market_df["Implied Odds"] = 1 / under_market_df["Prob"]
print("~  ~ under market ~  ~\n", f"{home_side} v. {away_side}\n", under_market_df)

~  ~ under market ~  ~
 Leicester v. Wolverhampton Wanderers
    Goals      Prob  Implied Odds
0    0.5  0.023701     42.192228
1    1.5  0.112396      8.897117
2    2.5  0.278355      3.592540
3    3.5  0.485373      2.060269


In [23]:
thresholds = [0.5, 1.5, 2.5, 3.5]
sums = []

for t in thresholds:
    mask = np.add.outer(
        np.arange(
            exact_score_prob.shape[0]
        ),
        np.arange(
            exact_score_prob.shape[1]
        )
    ) >= t
    sums.append(exact_score_prob[mask].sum())

# print(sums)
over_market_df = pd.DataFrame({
    "Goals": thresholds,
    "Prob": sums
})

over_market_df["Implied Odds"] = 1 / over_market_df["Prob"]
print("~  ~ over market ~  ~\n", f"{home_side} v. {away_side}\n", over_market_df)

~  ~ over market ~  ~
 Leicester v. Wolverhampton Wanderers
    Goals      Prob  Implied Odds
0    0.5  0.975872      1.024724
1    1.5  0.887177      1.127170
2    2.5  0.721219      1.386542
3    3.5  0.514200      1.944769


### 1 x 2 Probability

In [24]:
# Poisson probability for home team goals 0 - 8
home_score_prob = []
away_score_prob = []
for i in range(9):
    prob = poisson.pmf(i, home_projected_goals)
    home_score_prob.append(prob)
for i in range(9):
    prob = poisson.pmf(i, away_projected_goals)
    away_score_prob.append(prob)

home_score_prob = np.array(home_score_prob)
away_score_prob = np.array(away_score_prob)
# print(home_score_prob)
# print(away_score_prob)

exact_score_prob = np.outer(away_score_prob, home_score_prob)

# print(exact_score_prob)
home_win_prob = np.sum(np.triu(exact_score_prob, k=1))
away_win_prob = np.sum(np.tril(exact_score_prob, k=-1))
draw_prob = np.trace(exact_score_prob)

# prepare data in readable dataframe
data = {
    "winner": [f"{home_side} (1)", "Draw (x)", f"{away_side} (2)"],
    "prob": [home_win_prob, draw_prob, away_win_prob],
}
result_df = pd.DataFrame(data)
result_df['implied_odds'] = 1 / result_df['prob']
print("      ~  ~ moneyline market ~  ~\n", result_df)

      ~  ~ moneyline market ~  ~
                         winner      prob  implied_odds
0                Leicester (1)  0.499673      2.001310
1                     Draw (x)  0.207963      4.808537
2  Wolverhampton Wanderers (2)  0.291937      3.425393


# Full Results Summary:

In [25]:
# helper function to translate discrepency between Understat and Odds-Api naming
def translate(string):

    translations = {
        "Brighton": "Brighton and Hove Albion",
        "Ipswich": "Ipswich Town",
        "Leicester": "Leicester City",
        "Tottenham": "Tottenham Hotspur",
        "West Ham": "West Ham United"
    }

    if string in translations:
        return translations[string]
    return string

home_side, away_side = translate(home_side), translate(away_side)

match_string = f"{home_side} v. {away_side}"

print(f"CALCULATED PROBABILITIES FOR {home_side} v. {away_side}")
print("\n\n        ~ ~ moneyline market ~ ~\n\n", result_df)
print("\n\n          ~ ~ over market ~ ~ \n\n", over_market_df)
print("\n\n          ~ ~ under market ~ ~ \n\n", under_market_df)

try:
    odds_df = pd.read_csv(f"../data/processed/odds/odds_matchday_{matchday}.csv")
    filtered_odds = odds_df.loc[odds_df["match"] == match_string]
    implied = {
        "1": 1 / filtered_odds["1"],
        "x": 1 / filtered_odds["x"],
        "2": 1 / filtered_odds["2"],
        "point": filtered_odds["ovr_und_point"],
        "over": 1 / filtered_odds["over"],
        "under": 1 / filtered_odds["under"]
    }
    implied_odds_df = pd.DataFrame(implied)

    print(f"\n ~~ MyBookie Odds for {match_string} ~~\n",filtered_odds.drop(columns=["match"]))
    print(f"\n ~~ MyBookie Odds IMPLIED for {match_string} ~~\n", implied_odds_df)
except:
    print(f"No data for sportsbook odds on matchday {matchday}. To view against real odds, run `notebooks/03_retrieve_odds.ipynb` ran correctly and ensure matchday is correct in the first cell of this notebook")


CALCULATED PROBABILITIES FOR Leicester City v. Wolverhampton Wanderers


        ~ ~ moneyline market ~ ~

                         winner      prob  implied_odds
0                Leicester (1)  0.499673      2.001310
1                     Draw (x)  0.207963      4.808537
2  Wolverhampton Wanderers (2)  0.291937      3.425393


          ~ ~ over market ~ ~ 

    Goals      Prob  Implied Odds
0    0.5  0.975872      1.024724
1    1.5  0.887177      1.127170
2    2.5  0.721219      1.386542
3    3.5  0.514200      1.944769


          ~ ~ under market ~ ~ 

    Goals      Prob  Implied Odds
0    0.5  0.023701     42.192228
1    1.5  0.112396      8.897117
2    2.5  0.278355      3.592540
3    3.5  0.485373      2.060269

 ~~ MyBookie Odds for Leicester City v. Wolverhampton Wanderers ~~
       1     x     2  ovr_und_point  over  under
3  2.51  3.42  2.65            2.5  1.63   2.19

 ~~ MyBookie Odds IMPLIED for Leicester City v. Wolverhampton Wanderers ~~
           1         x        

# Conclusion and Pick
SAFE: Over 2.5
GUTS: Leicester to win at home. At 49% (ourmodel) there's value here. The lads are far from reliable though.

### My Pick:

